### ISSUE TITLE
Allow customization of @CacheFor cache key

### ISSUE THREAD

**Issue by tahseenarticle** [playframework/play1#1133]
At the moment when you annotate an Controller method with CacheFor. Play framework generate the cache key as follow in ActionInvoker.java line 152:

```
cacheKey = "urlcache:" + request.url + request.querystring;
```

CacheFor annotation does not allow customisation of this cache key. This is useful in case you your controller shows different page based on the session. For example if you viewing different language version of the website the URL will be the same but the locale information is in the session. Having a way to customise this cache key helps in lot different cases of caching.

My proposal is to extend the @CacheFor as follow:

```java
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface CacheFor {
    String value() default "1h";
    String id() default "";
    Class<? extends CacheKeyGenerator> generator() default DefaultCacheKeyGenerator.class;
}
```
Where CacheKeyGenerator is a simple interface that generate the cache key.

```java
public interface CacheKeyGenerator {
	public String generate(Request request);
}
```
```java
public class DefaultCacheKeyGenerator implements CacheKeyGenerator {
	@Override
	public String generate(Request request) {
		return "urlcache:" + request.url + request.querystring;
	}
}
```

Then we can modify the InvokerAction.invoke to generate the cache key based on the CacheKeyGenerator.

You can use a custom CacheKeyGenerator as follow:

```
@CacheFor(value = "1h", generator=SessionBasedCacheKeyGenerator.class)
```
If you do not specify the ```generator``` the default key generator will be used as usual.

```
@CacheFor(value = "1h")
```

PS: I am working on a pull request to this issue

**Comment by asolntsev** [comment]
@tahseenarticle Good luck! Let us know when the PR is ready.

**Comment by asolntsev** [comment]
@tahseenarticle Do you have some updates? Could you make a pull request? Can I help with this?

**Comment by asolntsev** [comment]
Implemented in PR #1166
